File: tests/fw/util/test_hdlc.c
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "clar.h"

#include "util/hdlc.h"

#include <string.h>

#include "stubs_passert.h"

// Setup

void test_hdlc__initialize(void) {
}

void test_hdlc__cleanup(void) {
}

// Tests

void test_hdlc__decode_no_special(void) {
  // without any special characters
  const char *str = "\x7eThis is a long string without any special characters to be escaped.\x7e";
  int len = strlen(str);
  HdlcStreamingContext ctx;
  hdlc_streaming_decode_reset(&ctx);
  for (int i = 0; i < len; i++) {
    char c = str[i];
    bool should_store, is_invalid;
    bool is_complete = hdlc_streaming_decode(&ctx, (uint8_t *)&c, &should_store, &is_invalid);
    cl_assert(is_invalid == false);
    if (i == 0 || i == len - 1) {
      cl_assert(is_complete == true);
      cl_assert(should_store == false);
    } else {
      cl_assert(is_complete == false);
      cl_assert(should_store == true);
      cl_assert(c == str[i]);
    }
  }
}

void test_hdlc__special_characters(void) {
  // make sure the escape characters haven't changed
  cl_assert(HDLC_FLAG == 0x7e);
  cl_assert(HDLC_ESCAPE == 0x7d);
  cl_assert(HDLC_ESCAPE_MASK == 0x20);
}

void test_hdlc__decode_empty(void) {
  // consecutive empty frames
  const uint8_t str[4] = {HDLC_FLAG, HDLC_FLAG, HDLC_FLAG, HDLC_FLAG};
  HdlcStreamingContext ctx;
  hdlc_streaming_decode_reset(&ctx);
  for (int i = 0; i < 4; i++) {
    char c = str[i];
    bool should_store, is_invalid;
    bool is_complete = hdlc_streaming_decode(&ctx, (uint8_t *)&c, &should_store, &is_invalid);
    cl_assert(is_complete == true);
    cl_assert(should_store == false);
    cl_assert(is_invalid == false);
  }
}

void test_hdlc__decode_invalid(void) {
  // invalid sequences
  uint8_t data;
  bool should_store, is_invalid, is_complete;
  HdlcStreamingContext ctx;

  // two consecutive escape characters
  hdlc_streaming_decode_reset(&ctx);
  data = HDLC_ESCAPE;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == false);
  cl_assert(is_invalid == false);
  data = HDLC_ESCAPE;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == false);
  cl_assert(is_invalid == true);

  // an escape character followed by a flag
  hdlc_streaming_decode_reset(&ctx);
  data = HDLC_ESCAPE;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == false);
  cl_assert(is_invalid == false);
  data = HDLC_FLAG;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == true);
  cl_assert(should_store == false);
  cl_assert(is_invalid == true);
}

void test_hdlc__decode_escaped_special(void) {
  // 2 escaped special characters
  uint8_t data;
  bool should_store, is_invalid, is_complete;
  HdlcStreamingContext ctx;
  hdlc_streaming_decode_reset(&ctx);

  // escaped escape character
  data = HDLC_ESCAPE;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == false);
  cl_assert(is_invalid == false);
  data = HDLC_ESCAPE ^ HDLC_ESCAPE_MASK;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == true);
  cl_assert(is_invalid == false);
  cl_assert(data == HDLC_ESCAPE);

  // escaped flag
  data = HDLC_ESCAPE;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == false);
  cl_assert(is_invalid == false);
  data = HDLC_FLAG ^ HDLC_ESCAPE_MASK;
  is_complete = hdlc_streaming_decode(&ctx, &data, &should_store, &is_invalid);
  cl_assert(is_complete == false);
  cl_assert(should_store == true);
  cl_assert(is_invalid == false);
  cl_assert(data == HDLC_FLAG);
}

void test_hdlc__encode_decode(void) {
  const char *str = "this is a string with the special \x7e \x7d \x7e\x7d \x7d\x7e characters";
  char buffer[100];
  int write_idx = 0;
  for (int i = 0; i < strlen(str); i++) {
    char c = str[i];
    if (hdlc_encode((uint8_t *)&c)) {
      buffer[write_idx++] = HDLC_ESCAPE;
    }
    buffer[write_idx++] = c;
  }
  buffer[write_idx++] = HDLC_FLAG;
  cl_assert(write_idx == strlen(str) + 7 /* 6 special characters to escape + 1 flag at end */);

  HdlcStreamingContext ctx;
  hdlc_streaming_decode_reset(&ctx);
  int read_idx = 0;
  int i = 0;
  while (true) {
    char c = buffer[i++];
    bool should_store, is_invalid;
    bool is_complete = hdlc_streaming_decode(&ctx, (uint8_t *)&c, &should_store, &is_invalid);
    cl_assert(is_invalid == false);
    if (should_store) {
      cl_assert(is_complete == false);
      cl_assert(c == str[read_idx++]);
    }
    if (is_complete) {
      cl_assert(should_store == false);
      cl_assert(i == write_idx);
      break;
    }
  }
  cl_assert(read_idx == strlen(str));
}

File: tools/tests/test_hdlc.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import sys
import tempfile
import unittest

# Allow us to run even if not at the `tools` directory.
root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
sys.path.insert(0, root_dir)

from hdlc import HDLCDecoder, hdlc_encode_data


class TestHDLCDecode(unittest.TestCase):
    def test_line_noise(self):
        decoder = HDLCDecoder()
        decoder.write(b'This is a bunch')
        decoder.write(b'of line noise!!\x7e')
        decoder.write(b'\x7eFollowed by a valid frame\x7e')
        self.assertEquals(decoder.get_frame(), 'Followed by a valid frame')
        self.assertIsNone(decoder.get_frame())

    def test_segmented_frames(self):
        decoder = HDLCDecoder()
        decoder.write(b'\x7eFrame ')
        decoder.write(b'one\x7eFrame two')
        decoder.write(b'\x7eFrame thr')
        decoder.write(b'ee\x7e')
        self.assertEquals(decoder.get_frame(), 'Frame one')
        self.assertEquals(decoder.get_frame(), 'Frame two')
        self.assertEquals(decoder.get_frame(), 'Frame three')
        self.assertIsNone(decoder.get_frame())

    def test_empty(self):
        decoder = HDLCDecoder()
        decoder.write(b'\x7e\x7e\x7e\x7e')
        self.assertIsNone(decoder.get_frame())

    def test_escape(self):
        decoder = HDLCDecoder()
        decoder.write(b'\x7eHow about escaping?\x7d\x5e\x7e')
        decoder.write(b'\x7eAny ch\x7dArac\x7dTer can be \x7dE\x7dS\x7dC\x7dA\x7dPed!\x7e')
        decoder.write(b'\x7eEven a \x7d\x7d\x7e')
        self.assertEquals(decoder.get_frame(), 'How about escaping?\x7e')
        self.assertEquals(decoder.get_frame(), 'Any character can be escaped!')
        self.assertEquals(decoder.get_frame(), 'Even a \x5d')
        self.assertIsNone(decoder.get_frame())

    def test_invalid(self):
        decoder = HDLCDecoder()
        decoder.write(b'\x7eInvalid termination\x7d\x7e')
        decoder.write(b'\x7eI am valid\x7e')
        decoder.write(b'partial frame')
        self.assertEquals(decoder.get_frame(), 'I am valid')
        self.assertIsNone(decoder.get_frame())


class TestHDLCEncodeData(unittest.TestCase):
    def test_simple(self):
        self.assertEquals(hdlc_encode_data('This is easy'), '\x7eThis is easy\x7e')

    def test_escape(self):
        self.assertEquals(hdlc_encode_data('Escape \x7d\x7e!'), '\x7eEscape \x7d\x5d\x7d\x5e!\x7e')

    def test_empty(self):
        self.assertEquals(hdlc_encode_data(''), '\x7e\x7e')


if __name__ == '__main__':
    unittest.main()
